<!-- dom:TITLE: Homework 1, PHY 959 -->
<!-- dom:AUTHOR: [PHY 959: Machine Learning in Physics]-->


# PHY 959: Homework Set #1
Due: **January 28, 2021**



___
***


# Problem 1: Getting comfortable with Python

In this problem, we will exercise some useful Python skills that will be useful in the rest of this homework problem set and for the rest of the course.  There will be three parts to this problem:

  * Loading the Iris Flower dataset.  Note that there are two questions to answer.
  * Making 1D histograms of the features 
  * Making 2D histograms of the features


## Part 1a:
We're going to be using Python3 (ideally v3.8!) so you should check that you have access to that version (go check now!).  There's no need to remove Python v2.X if you want to keep it, as your computer should allow both.  If you're here, I assume that you have anaconda already installed, so hopefully this part is done!

***

## Part 1b:
Next, we're going to be using NumPy, SciPy, scikit-learn, and Pandas this semester, so you'll want to ensure that you have them installed properly. If you're not familiar with pip, this is the primary installation tool for python.  Note: even if you have these, you should just do an upgrade to make sure you're using the same code as everyone else.  Double-note:  you may need to use pip3 for Python3.

` pip install -U numpy scipy pandas scikit-learn`


If you prefer to stay within anaconda, that's fine too:

`conda install -c anaconda numpy scipy pandas`

`conda install -c conda-forge scikit-learn`


As usual, report any failures or strange behavior to Prof. Fisher.

***

## Part 1c:

Now let's try to use our python skills to get a dataset!  The Iris Flowers Dataset is a labeled data sample that we can use for both supervised and unsupervised learning.  The code below will show you how to use pandas to load the Iris dataset:

In [6]:
# Load libraries
from pandas import read_csv

# Load dataset
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/iris.csv"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']
dataset = read_csv(url, names=names)



Now we can explore the data set by looking at the dimensionality, statistical summary and class description.


In [7]:
# summarize the data

# shape
print("Shape:\n",dataset.shape, "\n")

# first 20 samples
print("Head:\n",dataset.head(20), "\n")

# descriptions
print("Description:\n",dataset.describe(), "\n")

# class distributio
print("Class Distribution:\n",dataset.groupby('class').size())


Shape:
 (150, 5) 

Head:
     sepal-length  sepal-width  petal-length  petal-width        class
0            5.1          3.5           1.4          0.2  Iris-setosa
1            4.9          3.0           1.4          0.2  Iris-setosa
2            4.7          3.2           1.3          0.2  Iris-setosa
3            4.6          3.1           1.5          0.2  Iris-setosa
4            5.0          3.6           1.4          0.2  Iris-setosa
5            5.4          3.9           1.7          0.4  Iris-setosa
6            4.6          3.4           1.4          0.3  Iris-setosa
7            5.0          3.4           1.5          0.2  Iris-setosa
8            4.4          2.9           1.4          0.2  Iris-setosa
9            4.9          3.1           1.5          0.1  Iris-setosa
10           5.4          3.7           1.5          0.2  Iris-setosa
11           4.8          3.4           1.6          0.2  Iris-setosa
12           4.8          3.0           1.4          0.1  Iris-s


## Question #1:  
How many features are included in the Iris Flowers dataset?  Write your answer in the following cell.


## Question #2:  
What are the dimensions of the design matrix for this dataset (aka the feature matrix)?  Specify if you are including the labels in your answer.


***

## Part 2:

Now let's plot the data as histograms.  I suggest to use `matplotlib` module, but if you have other favorites that's fine too.  Make 1-dimensional histograms for each feature.  As you saw above, there are 150 entries so you should choose an appropriate number of histogram bins to get a good feeling for the PDF of each feature.  Calculate the mean and standard deviation of your histogram, comparing to the values found above.

Add cells below as needed.


In [8]:
#More modules!
from matplotlib import pyplot


***

## Part 3:

Last part!  Now let's make 2-dimensional figures for our data.  I suggest to use histograms for this.  How many 2D histograms should you make?

Hint:  I've started you off by loading the `scatter_matrix` module from Pandas.  Can you work out why?

In [9]:
#Hmmmm
from pandas.plotting import scatter_matrix
